ocamllib/src/cursorinfo/cursorInfo.ml

```diff
@@ -0,0 +1,424 @@
+      (def_cont : (Var.t * int) list) (typ_cont : Context.t)
+      (pred_type : Type.t option) (ind : int) : t =
+    let current : Syntax.t =
+      match node with
+      | ZENode zparent -> ENode (Expr.unzip_ast zparent)
+      | ZTNode zparent -> TNode (Type.unzip zparent)
+    in
+    match node with
+    | ZENode (Cursor cursed) ->
+        {
+          current_term = ENode cursed;
+          parent_term = parent;
+          ctx = def_cont;
+          typ_ctx = typ_cont;
+          expected_ty = pred_type;
+          actual_ty = synthesis typ_cont cursed;
+        }
+    | ZENode (EUnOp_L (OpNeg, argl)) ->
+        let exp_typ =
+          match pred_type with Some TBool | Some TInt -> pred_type | _ -> None
+        in
+        recurse (ZENode argl) (Some current) def_cont typ_cont exp_typ ind
+    | ZENode
+        (EBinOp_L
+          ( argl,
+            (OpPlus | OpMinus | OpTimes | OpDiv | OpGt | OpGe | OpLt | OpLe),
+            argr )) ->
+        recurse (ZENode argl) (Some current) def_cont typ_cont (Some TInt) ind
+    | ZENode
+        (EBinOp_R
+          ( argl,
+            (OpPlus | OpMinus | OpTimes | OpDiv | OpGt | OpGe | OpLt | OpLe),
+            argr )) ->
+        recurse (ZENode argr) (Some current) def_cont typ_cont (Some TInt)
+          (ind + Syntax.size (ENode argl) + 1)
+    | ZENode (EBinOp_L (argl, (OpEq | OpNe), argr)) ->
+        let exp_typ =
+          match pred_type with Some TBool | Some TInt -> pred_type | _ -> None
+        in
+        recurse (ZENode argl) (Some current) def_cont typ_cont exp_typ ind
+    | ZENode (EBinOp_R (argl, (OpEq | OpNe), argr)) ->
+        let exp_typ =
+          match pred_type with Some TBool | Some TInt -> pred_type | _ -> None
+        in
+        recurse (ZENode argr) (Some current) def_cont typ_cont exp_typ ind
+    | ZENode (EBinOp_L (argl, OpCons, arr)) ->
+        let exp_typ =
+          match (pred_type, synthesis typ_cont arr) with
+          | Some (TList ltype), _ -> Some ltype
+          | _, Some (TList ltype) -> Some ltype
+          | _ -> None
+        in
+        recurse (ZENode argl) (Some current) def_cont typ_cont exp_typ ind
+    | ZENode (EBinOp_R (argl, OpCons, arr)) ->
+        let exp_typ =
+          match (pred_type, synthesis typ_cont argl) with
+          | Some (TList ltype), _ -> Some ltype
+          | _, Some (TList ltype) -> Some ltype
+          | _ -> None
+        in
+        recurse (ZENode arr) (Some current) def_cont typ_cont exp_typ
+          (ind + Syntax.size (ENode argl) + 1)
+    | ZENode (EBinOp_L (argl, OpAp, argr)) -> (
+        match (pred_type, synthesis typ_cont argr) with
+        | Some out_type, Some inp_type ->
+            recurse (ZENode argl) (Some current) def_cont typ_cont
+              (Some (TArrow (inp_type, out_type)))
+              ind
+        | _ -> recurse (ZENode argl) (Some current) def_cont typ_cont None ind)
+    | ZENode (EBinOp_R (argl, OpAp, argr)) -> (
```

Comment:
expected type of argument should be: try to synthesize function term. if that synthesizes some_type->some_other_type, use some_type. in all other cases, THole